In [5]:
import pandas as pd
import os
import numpy as np

import tensorflow as tf
# pad_sequences
from keras.preprocessing.sequence import pad_sequences
# layers--->Embedding,LSTM,Dense,Bidirectional
from keras.layers import Embedding, LSTM, Dense, Bidirectional
# Tokenizer
from keras.preprocessing.text import Tokenizer
from keras.models import Sequential
from keras.optimizers import Adam

In [6]:
medium_data = pd.read_csv('medium_data.csv')
medium_data.head()

id  ...        date
0   1  ...  2019-05-30
1   2  ...  2019-05-30
2   3  ...  2019-05-30
3   4  ...  2019-05-30
4   5  ...  2019-05-30

[5 rows x 10 columns]

In [14]:
medium_data.shape

(6508, 10)

we have a 10 different fields and 6508 records but we will only use title field for predicting next word.

In [16]:
print("No of Records(row):",medium_data.shape[0])
print("No of fields(col):",medium_data.shape[1])

No of Records(row): 6508
No of fields(col): 10


In [17]:
medium_data['title'].head()

0    A Beginner’s Guide to Word Embedding with Gens...
1    Hands-on Graph Neural Networks with PyTorch & ...
2                         How to Use ggplot2 in Python
3    Databricks: How to Save Files in CSV on Your L...
4    A Step-by-Step Implementation of Gradient Desc...
Name: title, dtype: object

In [18]:
# Removing unwanted characters and words in titles
medium_data['title'] = medium_data['title'].apply(lambda x: x.replace(u'\xa0',u' '))
medium_data['title'] = medium_data['title'].apply(lambda x: x.replace('\u200a',' '))

In [19]:
# Tokenization
# Tokenization is the process in which we provide an unique id to all the words and make a word index or we can say vocabulary.

tokenizer = Tokenizer(oov_token='<oov>') # For those words which are not found in word_index
tokenizer.fit_on_texts(medium_data['title'])
total_words = len(tokenizer.word_index) + 1

print("Total number of words: ", total_words)
print("Word: ID")
print("------------")
print("<oov>: ", tokenizer.word_index['<oov>'])
print("Strong: ", tokenizer.word_index['strong'])
print("And: ", tokenizer.word_index['and'])
print("Consumption: ", tokenizer.word_index['consumption'])

Total number of words:  8238
Word: ID
------------
<oov>:  1
Strong:  4
And:  8
Consumption:  8237


In [21]:
input_sequences = []
for line in medium_data['title']:
    token_list = tokenizer.texts_to_sequences([line])[0]
    #print(token_list)
    for i in range(1, len(token_list)):
        n_gram_sequence = token_list[:i+1]
        input_sequences.append(n_gram_sequence)

# print(input_sequences)
print("Total input sequences: ", len(input_sequences))

Total input sequences:  48461


In [42]:
# Make all titles with same length by using padding
# The length of every title has to be the same.
# To make it, we need to find a title that has a maximum length, and based on that length, we have to pad rest of titles.
# pad sequences 
max_sequence_len = max([len(x) for x in input_sequences])
input_sequences = np.array(pad_sequences(input_sequences, maxlen=max_sequence_len, padding='pre'))
input_sequences[1]

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   5, 676,
        68])

In [44]:
# create features and label
xs, labels = input_sequences[:,:-1],input_sequences[:,-1]
ys = tf.keras.utils.to_categorical(labels, num_classes=total_words)

In [45]:
print(xs[5])
print(labels[5])
print(ys[5][14])

[   0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    5  676   68    2  452 1518]
14
1.0


BI Directional Lstm

In [51]:
model = Sequential()
model.add(Embedding(total_words, 100, input_length=max_sequence_len-1))
model.add(Bidirectional(LSTM(150)))
model.add(Dense(total_words, activation='softmax'))
adam = Adam(lr=0.01)
model.compile(loss='categorical_crossentropy', optimizer=adam, metrics=['accuracy'])
history = model.fit(xs, ys, epochs=1, verbose=1)
#print model.summary()
print(model)

c:\Users\hp\anaconda3\envs\pythonProject1\lib\site-packages\keras\optimizers\optimizer_v2\adam.py:114: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super().__init__(name, **kwargs)


1515/1515 [==============================] - 476s 312ms/step - loss: 6.7813 - accuracy: 0.1257


In [55]:
# import matplotlib.pyplot as plt


# def plot_graphs(history, string):
#     plt.plot(history.history[string])
#     plt.xlabel("Epochs")
#     plt.ylabel(string)
#     plt.show()

In [58]:
seed_text = "implementation of"
next_words = 2

for _ in range(next_words):
    token_list = tokenizer.texts_to_sequences([seed_text])[0]
    token_list = pad_sequences([token_list], maxlen=max_sequence_len-1, padding='pre')
    predicted_probabilities = model.predict(token_list, verbose=0)
    predicted_index = np.argmax(predicted_probabilities)
    output_word = ""
    for word, index in tokenizer.word_index.items():
        if index == predicted_index:
            output_word = word
            break
    seed_text += " " + output_word
print(seed_text)


implementation of the future
